### PR TITLE
Restore: improve performance by avoiding linear search in the implementation of the "Nearest wins" rule

### LINKED ISSUE #10286: Restore: improve performance by avoiding linear search in the implementation of the "Nearest wins" rule
## Details about Problem
I think that the current implementation of [nearest wins](https://docs.microsoft.com/en-us/nuget/concepts/dependency-resolution#nearest-wins ) rule can be improved so it has better performance.

Each time the graph walking algorithm is about to visit a child node (a dependency of a node) it checks whether that child node is not eclipsed or downgraded by any of the dependencies of parent nodes (direct and transitive). If a node is being ignored then all remaining dependencies on that branch of the graph are also ignored which is the key benefit of the "nearest wins" rule.
But, when we look closer at the [code](https://github.com/NuGet/NuGet.Client/blob/51266bcd0587e677e0bf942f99dd8c64f9de89ac/src/NuGet.Core/NuGet.DependencyResolver.Core/Remote/RemoteDependencyWalker.cs#L242) we can spot that it is implemented using linear search which is potentially bad for performance:
```
private Func<LibraryRange, (DependencyResult dependencyResult, LibraryDependency conflictingDependency)> ChainPredicate(Func<LibraryRange, (DependencyResult dependencyResult, LibraryDependency conflictingDependency)> predicate, GraphNode<RemoteResolveResult> node, LibraryDependency dependency)
{
	var item = node.Item;

	return library =>
	{
		if (StringComparer.OrdinalIgnoreCase.Equals(item.Data.Match.Library.Name, library.Name))
		{
			return (DependencyResult.Cycle, null);
		}

		//!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
		// LINEAR SEARCH
		//!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
		foreach (var d in item.Data.Dependencies)
		{
			if (d != dependency && library.IsEclipsedBy(d.LibraryRange))
			{
				if (d.LibraryRange.VersionRange != null &&
					library.VersionRange != null &&
					!IsGreaterThanOrEqualTo(d.LibraryRange.VersionRange, library.VersionRange))
				{
					return (DependencyResult.PotentiallyDowngraded, d);
				}

				return (DependencyResult.Eclipsed, d);
			}
		}

		return predicate(library);
	};
}
```
On the flip side, linear search is not always worse than a dictionary lookup - it all depends on the average number elements the linear search needs to go through. For narrow graphs, i.e. graphs where average number of dependencies of each node is low, the linear search can be equally fast or even outperform the dictionary lookup. Thus any change in this area needs to be well benchmarked using a representative set of solutions.

### PR DESCRIPTION
**Pull request by marcin-krystianc**
## Bug

Fixes: https://github.com/NuGet/Home/issues/10286
Regression: No

## Fix

The predicate returned by the `ChainPredicate` method has better algorithmic complexity than before. 
It uses dictionary lookup instead of linear search which as shown in the results below can have a significant impact on the overall restore time.

## Testing

Tests Added: No  
Reason for not adding tests: 
 - No new features are being added.
 - This part of code is already well tested

## Validation (Benchmarking)

Used solutions:
 - [NuGet.Client](https://github.com/NuGet/NuGet.Client)
 - [OrchardCore ](https://github.com/OrchardCMS/OrchardCore)
 - [Orleans](https://github.com/dotnet/orleans)
 - [SanitisedNet471](https://github.com/NuGet/Home/issues/10030#issue-701185886)
 
Test procedure:
- Vanilla [RunPerformanceTests.ps1](https://github.com/NuGet/NuGet.Client/blob/dev/scripts/perftests/RunPerformanceTests.ps1) script was used for running the individual benchmarks.
- As we are interested in the performance of the graph walking algorithm, the "force restore" scenario with static graph option was chosen.
- To achieve high confidence in the results each individual benchmark was run twice with high number of iterations (50 or 250)

Results:

|                        |  N-Runs |   Mean  |  Median |   Min   |   Max   | StdDev  |  Mean Change |
|------------------------|:-------:|:-------:|:-------:|:-------:|:-------:|:-------:|:-------:|
| NuGet.Client(Base)     |  2x250  |  16.57s |  16.57s |  15.72s |  17.57s |  0.320  |         |
| NuGet.Client(Branch)   |  2x250  |  16.46s |  16.47s |  15.46s |  17.49s |  0.314  |  -0.66% |
| Orleans(Base)          |  2x250  |  14.99s |  14.97s |  14.35s |  15.73s |  0.269  |         |
| Orleans(Branch)        |  2x250  |  14.88s |  14.87s |  13.93s |  15.49s |  0.265  |  -0.76% |
| OrchardCore(Base)      |  2x50   |  25.48s |  25.48s |  24.61s |  26.62s |  0.430  |         |
| OrchardCore(Branch)    |  2x50   |  23.84s |  23.84s |  22.82s |  25.03s |  0.471  |  -6.42% |
| SanitisedNet471(Base)  |  2x50   |  29.68s |  29.69s |  28.91s |  31.12s |  0.288  |         |
| SanitisedNet471(Branch)|  2x50   |  26.69s |  26.68s |  25.93s |  27.46s |  0.316  | -10.07% |

## Conclusions:
The improvement varies from negligible 0.5% up to even 10%, As it was explained in the original issue, it all depends on the average number elements the linear search needs to go through.
For solutions with narrow dependencies there is not so much of an improvement, but for solutions with wide dependencies it is really significant.

## Benchmark script & raw data
- [benchmark_dotnet.ps1](https://github.com/NuGet/NuGet.Client/files/5547424/benchmark_dotnet.ps1.txt)
- [results0_NuGet.Client.csv](https://github.com/NuGet/NuGet.Client/files/5547439/results0_NuGet.Client.csv.txt)
- [results1_NuGet.Client.csv](https://github.com/NuGet/NuGet.Client/files/5547443/results1_NuGet.Client.csv.txt)
- [results0_OrchardCore.csv](https://github.com/NuGet/NuGet.Client/files/5547440/results0_OrchardCore.csv.txt)
- [results1_OrchardCore.csv](https://github.com/NuGet/NuGet.Client/files/5547444/results1_OrchardCore.csv.txt)
- [results0_orleans.csv](https://github.com/NuGet/NuGet.Client/files/5547441/results0_orleans.csv.txt)
- [results1_orleans.csv](https://github.com/NuGet/NuGet.Client/files/5547445/results1_orleans.csv.txt)
- [results0_SanitisedNet471.csv](https://github.com/NuGet/NuGet.Client/files/5547442/results0_SanitisedNet471.csv.txt)
- [results1_SanitisedNet471.csv](https://github.com/NuGet/NuGet.Client/files/5547446/results1_SanitisedNet471.csv.txt)

### CODE REVIEW

File: test/NuGet.Core.Tests/NuGet.Commands.Test/RestoreCommandTests.cs [hunk 1]
```diff
@@ -1542,10 +1543,6 @@ public async Task RestoreCommand_LogDowngradeWarningsOrErrorsAsync_ErrorWhenCpvm
             provider.Package("C", otherVersion)
                   .DependsOn(centralPackageName, otherVersion);
 
-            // Simulates the existence of a D centrally defined package that is not direct dependency
```
Comment (marcin-krystianc) [inline test/NuGet.Core.Tests/NuGet.Commands.Test/RestoreCommandTests.cs:1548]: This was a duplicate of line 1537

File: src/NuGet.Core/NuGet.DependencyResolver.Core/Remote/RemoteDependencyWalker.cs [hunk 2]
```diff
@@ -239,30 +247,30 @@ public async Task<GraphNode<RemoteResolveResult>> WalkAsync(LibraryRange library
...
             return library =>
             {
-                if (StringComparer.OrdinalIgnoreCase.Equals(item.Data.Match.Library.Name, library.Name))
+                if (StringComparer.OrdinalIgnoreCase.Equals(nodeName, library.Name))
                 {
                     return (DependencyResult.Cycle, null);
                 }
 
-                foreach (var d in item.Data.Dependencies)
+                if (dependencies.TryGetValue(library.Name, out var d) &&
```
Comment (erdembayar) [inline src/NuGet.Core/NuGet.DependencyResolver.Core/Remote/RemoteDependencyWalker.cs:262]: I believe this one will introduce correctness issue. Before logic was at each child node we step back into parent then cross check for cycle or downgrade problem at other child dependency of parent, same time repeatedly go up to another level parent to exhaust of cases. But now it looks we're only looking at 1 branch of code execution not exhaustive as before. 
If we have extensive coverage unit test for this most likely this one cause failure.

Comment (erdembayar) [reply src/NuGet.Core/NuGet.DependencyResolver.Core/Remote/RemoteDependencyWalker.cs:262]: Also even though it looks expensive usually C# code is still very fast compare to IO read/write, reflection etc...
So I'm not really sure if we can improve with only simple dictionary, maybe can do large dictionary that can lookup not only library/dependency also has information about which level it was seen, so while we're going top to bottom no need to go up and check with parents. It'll be 1 pass algorithm instead of recursion, but problem is memory might balloon up.

Comment (marcin-krystianc) [reply src/NuGet.Core/NuGet.DependencyResolver.Core/Remote/RemoteDependencyWalker.cs:262]: > Also even though it looks expensive usually C# code is still very fast compare to IO read/write, reflection etc...

Generally you are right but in this case my change actually improves performance a lot. See the performance results. For some solutions even this simple dictionary lookup improves the restore time by 10%! Predicates created with `ChainPredicate` are on the critical path of the restore graph construction, so any micro-optimisation here can be worth doing.

> So I'm not really sure if we can improve with only simple dictionary, maybe can do large dictionary that can lookup not only library/dependency also has information about which level it was seen, so while we're going top to bottom no need to go up and check with parents. It'll be 1 pass algorithm instead of recursion, but problem is memory might balloon up.

Your idea about getting rid of recursion seems interesting, I think it can make it even faster. I'll give it a try.

Comment (marcin-krystianc) [reply src/NuGet.Core/NuGet.DependencyResolver.Core/Remote/RemoteDependencyWalker.cs:262]: > I believe this one will introduce correctness issue. Before logic was at each child node we step back into parent then cross check for cycle or downgrade problem at other child dependency of parent, same time repeatedly go up to another level parent to exhaust of cases. But now it looks we're only looking at 1 branch of code execution not exhaustive as before.
> If we have extensive coverage unit test for this most likely this one cause failure.

I don't think that my modification has changed how it works. It is still going to cross check given node with its parent nodes for cycles and direct dependencies of parent nodes for version eclipsing and downgrades.

**Comment by zivkan** [comment]
@marcin-krystianc Did you compare assets files of NuGet.Client, Orleans, and/or Orchard.Core projects with and without your changes, to ensure the results are really the same?

I don't have an enormous amount of trust in our unit/functional tests. Some trust yes, but if the resolved graph ends up being different, a whole lot of people might be broken. Even when the test authors try to write tests for all the edge cases they can think of, they usually involve a small number of packages and shallow dependency graph. I wouldn't be surprised if there are scenarios that are untested.

File: src/NuGet.Core/NuGet.DependencyResolver.Core/Remote/RemoteDependencyWalker.cs (hunk 2 above)
Comment (marcin-krystianc) [reply src/NuGet.Core/NuGet.DependencyResolver.Core/Remote/RemoteDependencyWalker.cs:262]: > So I'm not really sure if we can improve with only simple dictionary, maybe can do large dictionary that can lookup not only library/dependency also has information about which level it was seen, so while we're going top to bottom no need to go up and check with parents. It'll be 1 pass algorithm instead of recursion, but problem is memory might balloon up.

To get rid of recursion to parents I ended up with one dictionary for checking dependencies and one hash set for checking cycles. Unfortunately that has made things even slower than the original linear search. I even tried using immutable data structures to reduce gc pressure (e.g. using `ImmutableHashSet` for cycle detection) but that didn't help either. 
I think that the current approach of replacing linear search with simple dictionary lookup and chaining predicates (so we need to use recursion to parents) is the optimal one. 
I believe that constructing a large dictionary containing dependencies from all levels repeats the work that was done already on previous levels. The time that is saved on dictionary lookups is lost on more expensive dictionary construction and more expensive GC (larger dictionaries mean more memory turnover).

Comment (erdembayar) [reply src/NuGet.Core/NuGet.DependencyResolver.Core/Remote/RemoteDependencyWalker.cs:262]: > > So I'm not really sure if we can improve with only simple dictionary, maybe can do large dictionary that can lookup not only library/dependency also has information about which level it was seen, so while we're going top to bottom no need to go up and check with parents. It'll be 1 pass algorithm instead of recursion, but problem is memory might balloon up.
> 
> To get rid of recursion to parents I ended up with one dictionary for checking dependencies and one hash set for checking cycles. Unfortunately that has made things even slower than the original linear search. I even tried using immutable data structures to reduce gc pressure (e.g. using `ImmutableHashSet` for cycle detection) but that didn't help either.
> I think that the current approach of replacing linear search with simple dictionary lookup and chaining predicates (so we need to use recursion to parents) is the optimal one.
> I believe that constructing a large dictionary containing dependencies from all levels repeats the work that was done already on previous levels. The time that is saved on dictionary lookups is lost on more expensive dictionary construction and more expensive GC (larger dictionaries mean more memory turnover).

Could be able share your code maybe in another branch or with this PR, I want to see how your dictionary structure, I'm wondering what happened. Many cases runtime complexity can be solved with more memory but sometimes we can reduce both of them. Also include if you have benchmarking code.

Comment (marcin-krystianc) [reply src/NuGet.Core/NuGet.DependencyResolver.Core/Remote/RemoteDependencyWalker.cs:262]: > Could be able share your code maybe in another branch or with this PR, I want to see how your dictionary structure, I'm wondering what happened. Many cases runtime complexity can be solved with more memory but sometimes we can reduce both of them. Also include if you have benchmarking code.

I've created another branch to avoid clobbering this PR: https://github.com/NuGet/NuGet.Client/compare/dev...marcin-krystianc:marcink-20201118-nochainpredicate

The benchmarking procedure is explained in the description of this PR. In short I was preparing custom dotnet sdks (using release configuration) and then I was using https://github.com/NuGet/NuGet.Client/files/5547424/benchmark_dotnet.ps1.txt to run the benchmark across different solutions.

**Comment by marcin-krystianc** [comment]
> @marcin-krystianc Did you compare assets files of NuGet.Client, Orleans, and/or Orchard.Core projects with and without your changes, to ensure the results are really the same?
> 
> I don't have an enormous amount of trust in our unit/functional tests. Some trust yes, but if the resolved graph ends up being different, a whole lot of people might be broken. Even when the test authors try to write tests for all the edge cases they can think of, they usually involve a small number of packages and shallow dependency graph. I wouldn't be surprised if there are scenarios that are untested.

Yes I did compare `project.assets.json` files for `NuGet.Client`, `Orleans`, and `Orchard.Core` today. Apart from the `runtimeIdentifierGraphPath` field, there were no differences.

**Comment by erdembayar** [comment]
> Yes I did compare `project.assets.json` files for `NuGet.Client`, `Orleans`, and `Orchard.Core` today. Apart from the `runtimeIdentifierGraphPath` field, there were no differences.

Can you investigate why that `runtimeIdentifierGraphPath` field is different? I'll review your other branch with dictionary when I have time.

**Comment by nkolev92** [comment]
> Yes I did compare project.assets.json files for NuGet.Client, Orleans, and Orchard.Core today. Apart from the runtimeIdentifierGraphPath field, there were no differences.

The rid graph path is set by the SDK, so it's driven by whichever SDK is in use for that project. 

Did you test `nuget.exe` or something else?

**Comment by erdembayar** [comment]
> > Yes I did compare project.assets.json files for NuGet.Client, Orleans, and Orchard.Core today. Apart from the runtimeIdentifierGraphPath field, there were no differences.
> 
> The rid graph path is set by the SDK, so it's driven by whichever SDK is in use for that project.
> 
> Did you test `nuget.exe` or something else?

But isn't same project/solution use same SDK regardless of this change?

**Comment by nkolev92** [comment]
> But isn't same project/solution use same SDK regardless of this change?

Not if you test with `dotnet.exe`. Then it's the version is picked through that.

**Comment by marcin-krystianc** [comment]
> Did you test `nuget.exe` or something else?

I just did the "find & replace job" job to update the RID graph paths. After updating them, there were no other differences in these assets files.

**Comment by nkolev92** [comment]
> I just did the "find & replace job" job to update the RID graph paths. After updating them, there were no other differences in these assets files.

The SDK being used can make a difference in performance. Especially if `dotnet restore` was being used. 


Were you testing with nuget.exe?

**Comment by marcin-krystianc** [comment]
> The SDK being used can make a difference in performance. Especially if `dotnet restore` was being used.
> 
> Were you testing with nuget.exe?

I made sure that the only difference between sdks was in `NuGet.*.dll` files that were compiled by me (from `dev` branch and from my branch). So I believe that differences in results are caused only by changes on my branch.

To clear any doubts I performed the same benchmark using `NuGet.exe` compiled from these branches (with `.\build.ps1 -SkipUnitTest -Configuration Release` command).

### Results
| NuGet.exe  restore -force |  N-Runs |   Mean  |  Median |   Min   |   Max   | StdDev  |  Mean Change |
|------------------------|:-------:|:-------:|:-------:|:-------:|:-------:|:-------:|:-------:|
| NuGet.Client(Base)     |  2x250  |  26.43s |  26.42s |  25.93s |  29.22s |  0.241  |         |
| NuGet.Client(Branch)   |  2x250  |  26.41s |  26.39s |  25.97s |  28.05s |  0.214  |  -0.08% |
| Orleans(Base)          |  2x250  |  15.13s |  15.12s |  14.89s |  16.47s |  0.105  |         |
| Orleans(Branch)        |  2x250  |  15.05s |  15.04s |  14.83s |  15.66s |  0.082  |  -0.52% |
| OrchardCore(Base)      |  2x250  |  27.63s |  27.59s |  27.19s |  31.60s |  0.303  |         |
| OrchardCore(Branch)    |  2x250  |  26.76s |  26.74s |  26.23s |  27.76s |  0.240  |  -3.17% |
| SanitisedNet471(Base)  |  2x250  |  39.16s |  39.12s |  38.59s |  40.14s |  0.287  |         |
| SanitisedNet471(Branch)|  2x250  |  36.84s |  36.84s |  36.03s |  38.98s |  0.256  |  -5.91% |

### Benchmark script & raw data
- [benchmark.ps1](https://github.com/NuGet/NuGet.Client/files/5596544/benchmark.ps1.txt)
- [results0_NuGet.Client.csv](https://github.com/NuGet/NuGet.Client/files/5596545/results0_NuGet.Client.csv.txt)
- [results1_NuGet.Client.csv](https://github.com/NuGet/NuGet.Client/files/5596549/results1_NuGet.Client.csv.txt)
- [results0_orleans.csv](https://github.com/NuGet/NuGet.Client/files/5596551/results0_orleans.csv.txt)
- [results1_orleans.csv](https://github.com/NuGet/NuGet.Client/files/5596552/results1_orleans.csv.txt)
- [results0_OrchardCore.csv](https://github.com/NuGet/NuGet.Client/files/5596554/results0_OrchardCore.csv.txt)
- [results1_OrchardCore.csv](https://github.com/NuGet/NuGet.Client/files/5596555/results1_OrchardCore.csv.txt)
- [results0_SanitisedNet471.csv](https://github.com/NuGet/NuGet.Client/files/5596556/results0_SanitisedNet471.csv.txt)
- [results1_SanitisedNet471.csv](https://github.com/NuGet/NuGet.Client/files/5596559/results1_SanitisedNet471.csv.txt)

I also compared assets files for both `NuGet.exe` versions (for all benchmarked solutions) - and this time there were absolutely no differences at all.

### Conclusions
Changes on this PR improve restore time for both `dotnet` and `NuGet.exe`. For some of the tested solutions, the difference is negligible but for others it is rather significant.

**Comment by marcin-krystianc** [comment]
Hi, is there anything I can do to help reviewing this PR?

**Review by nkolev92 — COMMENTED** [review]
Hey @marcin-krystianc 

Thanks for this contribution and apologies for the long overdue review.

The change looks logically same, so I don't have correctness concerns. The description did confuse me a bit, given that this only affects the backtracking pass to find the downgrades more than anything else. 

Can you please rebase so I can run the tests on the CI?

I'll try to run some of our perf tests to confirm your findings.

**Comment by marcin-krystianc** [comment]
> Thanks for this contribution and apologies for the long overdue review.

No worries.

> Can you please rebase so I can run the tests on the CI?

Merged the latest `dev` branch and resolved conflicts. is that now ok for CI ?

**Comment by nkolev92** [comment]
Hey @marcin-krystianc 

Apologies for the extremely low overdue review.
We have been [extremely busy](https://github.com/NuGet/Home/pull/10660) with the package namespaces work and totally dropped the ball here. We will try to be more proactive. 

I took a dev baseline alongside dev + your change and measured on a few different machines(different specs), along with a few real world solution and I was not able to observe a notable improvement consistently in our real world solutions (I used NuGet.Client, SDK repo, orleans, OrchardCore). 
Note that I ran tests with `MSBuild.exe` and that has slightly different performance characteristics than dotnet.exe, because of the different runtime. 

As you have observed the problem is particularly apparent in graphs with many dependencies. 
According to nuget.org, listed packages, on average, have 3.5 dependencies (if you include packages without deps it goes down to 2.5). 
I did some benchmarking of the performance of linear lookup vs dictionary and the I observed that the benefit manifests at 7 elements.

One obvious problem in the current implementation is that the `search` is exhaustive. That is a remnant of early project.json designs where NuGet restore accounted for way more than just package references. 

As such I think the better approach would be to not pay the allocation cost and just stop when the first id that matches the node being inspected.

**Comment by marcin-krystianc** [comment]
@nkolev92 Thanks for looking at this.

Let me run again performance test using different clients (`msbuild`, `dotnet` and `nuget.exe` ). I'll also try to find more examples of solutions for which my change makes a difference.  I think also that I can avoid some allocations by using a pool and by not using LINQ.

> One obvious problem in the current implementation is that the search is exhaustive. That is a remnant of early project.json designs where NuGet restore accounted for way more than just package references. As such I think the better approach would be to not pay the allocation cost and just stop when the first id that matches the node being inspected.

Can you elaborate on this ? I think that the current implementation on the `dev` does exactly that, it stops when it encounters any id matching the current node. I don't see how this could be simplified.

**Comment by nkolev92** [comment]
> Can you elaborate on this ? I think that the current implementation on the dev does exactly that, it stops when it encounters any id matching the current node. I don't see how this could be simplified.

@marcin-krystianc I misread it, my bad, it stops at the first hit.

**Comment by nkolev92** [comment]
Hey @marcin-krystianc 

I wanted to take you for all the work on this PR. 
The current version of the PR does not really move the needle in the benchmarks that we performed, so I'd like to close it at this point. 

We'd be happy to continue the discussion as needed.

**Comment by marcin-krystianc** [comment]
@nkolev92 I have a hunch that this PR can still make a significant difference when we merge the PR which enables transitive dependency pinning (#4025). With transitive dependency pinning, the dependency graph can get much wider, so it might be very beneficial to do dictionary lookup instead of a linear scan. I'm going to run some benchmarks to see whether I'm right and will keep you posted.

**Comment by marcin-krystianc** [comment]
@nkolev92 It is only partially true that transitive dependency pinning leads to a wide dependency graph, it only happens for the top-level nodes and all other nodes remain the same. So as a result a linear scan seems to be good enough in such cases (confirmed experimentally). 
However, I've managed to create an example of a solution without relying on transitive dependency pinning, where a linear scan performs **5-fold** worse than a dictionary lookup.  It is a solution that combines the problem of inefficiency of linear scans with the problem of exponential complexity of the restore algorithm (https://github.com/NuGet/Home/issues/10030).
- Link to a solution: [ChainPredicateBomb](https://github.com/marcin-krystianc/TestSolutions/tree/master/ChainPredicateBomb)
- Average restore time using linear scan: 100s
- Average restore time using dictionary lookup: 20s

Although it was possible to prepare a solution that exposes performance problems with the linear scan, I agree that in real-world scenarios the problem with the linear scan seems to be a non-issue. At this point, it seems to me that https://github.com/NuGet/Home/issues/10030 is more important to fix as it actually hinders our productivity. Also, I have a hunch that fixing https://github.com/NuGet/Home/issues/10030 will also make https://github.com/NuGet/Home/issues/10286 unexploitable anyway.
